test/jdk/java/net/httpclient/FileChannelPublisherTest.java

```diff
@@ -0,0 +1,693 @@
+            throw new UncheckedIOException(exception);
+        }
+    }
+
+    private record ServerRequestPair(
+            String serverName,
+            HttpTestServer server,
+            BlockingQueue<byte[]> serverReadRequestBodyBytes,
+            HttpRequest.Builder requestBuilder,
+            boolean secure) {
+
+        private static CountDownLatch SERVER_REQUEST_RECEIVED_SIGNAL = null;
+
+        private static CountDownLatch SERVER_READ_PERMISSION = null;
+
+        private static ServerRequestPair of(Version version, boolean secure) {
+
+            // Create the server
+            SSLContext sslContext = secure ? SSL_CONTEXT : null;
+            HttpTestServer server = createServer(version, sslContext);
+            String serverName = secure ? version.toString().replaceFirst("_", "S_") : version.toString();
+
+            // Add the handler
+            String handlerPath = "/%s/".formatted(CLASS_NAME);
+            BlockingQueue<byte[]> serverReadRequestBodyBytes =
+                    addRequestBodyConsumingServerHandler(serverName, server, handlerPath);
+
+            // Create the request builder
+            String requestUriScheme = secure ? "https" : "http";
+            // `x` suffix in the URI is not a typo, but ensures that *only* the parent handler path is matched
+            URI requestUri = URI.create("%s://%s%sx".formatted(requestUriScheme, server.serverAuthority(), handlerPath));
+            HttpRequest.Builder requestBuilder = HttpRequest.newBuilder(requestUri).version(version);
+
+            // Create the pair
+            ServerRequestPair pair = new ServerRequestPair(serverName, server, serverReadRequestBodyBytes, requestBuilder, secure);
+            pair.server.start();
+            LOGGER.log("Server[%s] is started at `%s`", pair, server.serverAuthority());
+
+            return pair;
+
+        }
+
+        private static HttpTestServer createServer(Version version, SSLContext sslContext) {
+            try {
+                // The default HTTP/1.1 test server processes requests sequentially.
+                // This causes a deadlock for concurrent tests such as `testSlicedUpload()`.
+                // Hence, explicitly providing a multithreaded executor for HTTP/1.1.
+                ExecutorService executor = Version.HTTP_1_1.equals(version) ? EXECUTOR : null;
+                return HttpTestServer.create(version, sslContext, executor);
+            } catch (IOException ioe) {
+                throw new UncheckedIOException(ioe);
+            }
+        }
+
+        private static BlockingQueue<byte[]> addRequestBodyConsumingServerHandler(
+                String serverName, HttpTestServer server, String handlerPath) {
+            BlockingQueue<byte[]> readRequestBodyBytes = new LinkedBlockingQueue<>();
+            HttpTestHandler handler = exchange -> {
+                // `HttpTestExchange::toString` changes on failure, pin it
+                String exchangeName = exchange.toString();
```

Comment:
Nit - the use of exchange name may not be necessary and instead we could maybe just use the request URI (`getRequestURI()`) in the log messages.

Reply:
Dumping the exchange name (which includes the object identifier!) has proved in several troubleshooting occasions very helpful in this test, in particular, when request handler output can be interleaved due to either timing issues in sequential requests or concurrent requests. I settled on this version after trying out the request URI, which did not help much since it is _always_ the same.